pyat/at/physics/rdt.py

```diff
@@ -0,0 +1,575 @@
+    h11110: Sequence[complex] | complex = None
+    h00220: Sequence[complex] | complex = None
+    h31000: Sequence[complex] | complex = None
+    h40000: Sequence[complex] | complex = None
+    h20110: Sequence[complex] | complex = None
+    h11200: Sequence[complex] | complex = None
+    h20020: Sequence[complex] | complex = None
+    h20200: Sequence[complex] | complex = None
+    h00310: Sequence[complex] | complex = None
+    h00400: Sequence[complex] | complex = None
+    # RDTtype.DETUNING
+    dnux_dJx: Sequence[float] | float = None
+    dnux_dJy: Sequence[float] | float = None
+    dnuy_dJy: Sequence[float] | float = None
+
+    def __getattr__(self, item):
+        return asdict(self)[item]
+
+    def __getitem__(self, item):
+        return asdict(self)[item]
+
+
+def _get_polynom(elem, attr, index):
+    try:
+        val = getattr(elem, attr)[index]
+        return val
+    except (IndexError, AttributeError):
+        return 0
+
+
+def _compute_pf(tune, nperiods):
+    "This uses the formula Sum(x^k, k=1->p) = x(x^p-1)/(x-1)"
+    if nperiods != 1:
+        for i in range(9):
+            for j in range(9):
+                a1 = np.pi * 2 * (tune[0] * (i - 4) + tune[1] * (j - 4))
+                a2 = a1 / nperiods
+                _PERIODICFACTOR[i][j] = (np.exp(1j * a1) - 1.0) / (
+                    np.exp(1j * a2) - 1.0
+                )
+
+
+def _computedrivingterms(
+    s,
+    betax,
+    betay,
+    phix,
+    phiy,
+    etax,
+    a2l,
+    b2l,
+    b3l,
+    b4l,
+    tune,
+    rdttype,
+    nperiods,
+    refpt,
+    second_order,
+):
+    """
+    Original implementation from ELEGANT
+    Based on J.Bengtsson, SLS Note 9 / 97, March 7, 1997, with corrections per W.Guo (NSLS)
+    Revised to follow C.X.Wang AOP - TN - 2009 - 020 for second - order terms
+    A. Franchi et al. arxiv 1711.06589, PRAB 17.074001
+    """
+
+    def pf(i, j):
+        return _PERIODICFACTOR[4 + i][4 + j]
+
+    rdts = {}
+    rbetax = np.sqrt(betax)
+    rbetay = np.sqrt(betay)
+    px = np.exp(1j * phix)
+    py = np.exp(1j * phiy)
+
+    rdts["refpts"] = refpt
+    mask_a2l = np.absolute(a2l) > 1.0e-6
+    mask_b2l = np.absolute(b2l) > 1.0e-6
+    mask_b3l = np.absolute(b3l) > 1.0e-6
+    mask_b4l = np.absolute(b4l) > 1.0e-6
+
+    if (RDTType.FOCUSING in rdttype) or (RDTType.ALL in rdttype):
+        b2lm = b2l[mask_b2l]
+        betaxm = betax[mask_b2l]
+        betaym = betay[mask_b2l]
+        pxm = px[mask_b2l]
+        pym = py[mask_b2l]
+        rdts["h20000"] = sum((b2lm / 8) * betaxm * pxm * pxm * pf(2, 0))
+        rdts["h00200"] = -sum((b2lm / 8) * betaym * pym * pym * pf(0, 2))
+
+    if (RDTType.COUPLING in rdttype) or (RDTType.ALL in rdttype):
+        a2lm = a2l[mask_a2l]
+        rbetaxm = rbetax[mask_a2l]
+        rbetaym = rbetay[mask_a2l]
+        pxm = px[mask_a2l]
+        pym = py[mask_a2l]
+        rdts["h10010"] = sum((a2lm / 4) * rbetaxm * rbetaym * pxm / pym * pf(1, -1))
```

Comment:
There should be a minus sign in front of the sum.
Here is the explanation:
From Bengtsson's Eq(15) $H\propto-A_s$.
In Eq(63) $M_{\xi0\rightarrow1} = \exp(-\int H ds$).
For a thin lens magnet this implies $M_{\xi0\rightarrow1} = exp(LA_s)$, which corresponds to the expansion terms in Eq(71).
Therefore, for the kth element $V_i = L_kA_{s,k}$.
Using Eq(8) for a skew quad $V_i=-L_k\mathbb{Re}\frac{ia_2}{2}(x+iy)^2=a_2xyL_i$.
Using the resonant base in Eq(87) and applying the rotation and amplitude transformations as in Eq(72) to find $\hat V_i$, we get
$a_2L_k\sqrt{\beta_x\beta_y}(\frac{1}{2}h_x^+e^{i\mu_x}+\frac{1}{2}h_x^- e^{-i\mu_x})(\frac{1}{2}h_y^+e^{i\mu_y}+\frac{1}{2}h_y^- e^{-i\mu_y}).$
We take only the term for `h10010`.
$h_{10010}=\frac{1}{4}a_2L_k\sqrt{\beta_x\beta_y}e^{i\mu_x}e^{-i\mu_y}$.
Bengtsson defines the magnetic field in Eq(10 ) with opposite sign wrt AT. Therefore,
$h_{10010,AT}=-\frac{1}{4}a_2L_k\sqrt{\beta_x\beta_y}e^{i\mu_x}e^{-i\mu_y}$.